### ISSUE TITLE
a proper way to do gradient clipping in the first 1 or 2 epochs when train a detector?

### ISSUE THREAD

**Issue by zimenglan-sysu-512** [facebookresearch/maskrcnn-benchmark#347]
## ❓ Questions and Help
is there a proper way to do gradient clipping in the first 1 or 2 epochs when train a detector? since training my own dataset sometimes occurs the NaN.

**Comment by fmassa** [comment]
You could use `torch.nn.utils.clip_grad_norm_` or `torch.nn.utils.clip_grad_value_`, see https://pytorch.org/docs/stable/nn.html?highlight=clip#torch.nn.utils.clip_grad_norm_

This is currently not implemented in the codebase, could easily be added by modifying the trainer and adding the clip just before the `optimizer.step` in https://github.com/facebookresearch/maskrcnn-benchmark/blob/d28845e112de36781b2b5f7217a34b2b62de8d2f/maskrcnn_benchmark/engine/trainer.py#L76-L77

I'm closing this issue, but let me know if you have further questions.

**Comment by zimenglan-sysu-512** [comment]
thanks @fmassa

**Comment by zimenglan-sysu-512** [comment]
hi @fmassa 
as u mention "adding the clip just before the optimizer.step", does it clip the grads before reducing the grads?
how to clip the grads after reducing the grads?
thansk.
